# CSC 593

## Week 7

### Merge Errors

#### Not resolved automatically
`git checkout --ours PATH/FILE`

#### Resolved automatically (vim)
`:wq`

### Pandas 2

In [1]:
import numpy as np
import pandas as pd

The BRFSS (Behavioral Risk Factor Surveillance System) data is too big to put into Github. This cell downloads it from the CDC's website and unzips it into your `data` folder.

In [2]:
#Setup for examples.
from urllib.request import urlretrieve
import zipfile
from pathlib import Path

zf = '../data/brfss/LLCP2018ASC.zip'
if not Path(zf).exists():
    Path('../data/brfss').mkdir(exist_ok=True)

    urlretrieve('https://www.cdc.gov/brfss/annual_data/2018/pdf/overview-2018-508.pdf', '../data/brfss/overview-2018-508.pdf')
    urlretrieve('https://www.cdc.gov/brfss/annual_data/2018/pdf/codebook18_llcp-v2-508.pdf', '../data/brfss/codebook18_llcp-v2-508.pdf')
    
    urlretrieve('https://www.cdc.gov/brfss/annual_data/2018/files/LLCP2018ASC.zip', zf)

fwff = '../data/brfss/LLCP2018.ASC'
if not Path(fwff).exists():
    with zipfile.ZipFile(zf) as z:
        z.extractall('../data/brfss')

Load the BRFSS data and set a couple of data types explicitly. (More supported data types are listed at https://docs.scipy.org/doc/numpy/user/basics.types.html)

In [4]:
names= ['state', 'imonth', 'iday', 
        'iyear', 'dispcode','genhlth', 
        'physhlth',
        'menthlth', 'poorhlth', 'hlthpln1',
        'persdoc2', 'medcost', 'checkup1',
        'WEIGHT2', 'HEIGHT3']
cols = [
    (1, 3),
    (18, 20),
    (20, 22),
    (22, 27),
    (31, 35),
    (89, 90),
    (90, 92),
    (92, 94),
    (94, 96),
    (96, 97),
    (97, 98),
    (98, 99),
    (99, 100),
    (176, 180),
    (180, 184)
]
types= {
    'WEIGHT2': str, 
    'HEIGHT3': str,
}
brfss = pd.read_fwf(fwff + ' ', 
                    names=names,
                    colspecs=cols,
                    dtype=types)

#### Searching

In [5]:
#Get an individual column.
brfss['WEIGHT2']

0         0130
1         0200
2         0142
3         0190
4         7777
5         0172
6         0150
7         0130
8         0205
9         0151
10        0199
11        0170
12        0160
13        0200
14        0285
15        0217
16        0230
17        0250
18        0155
19        0138
20        0240
21        0240
22        0185
23        0165
24        0131
25        0119
26        0180
27        0169
28        0128
29        0218
          ... 
437406     NaN
437407    0320
437408    0125
437409    0158
437410    0155
437411    0150
437412    0138
437413    0135
437414    0178
437415    0126
437416    0115
437417    0183
437418    0153
437419    0138
437420    0260
437421    0160
437422    0200
437423    0126
437424    0340
437425    0298
437426    0160
437427    0160
437428    0163
437429    0172
437430    0170
437431    0098
437432    7777
437433    0156
437434    0137
437435    0192
Name: WEIGHT2, Length: 437436, dtype: object

In [6]:
#For multiple columns, use a list as a subscript.
brfss[['WEIGHT2', 'HEIGHT3']]

WEIGHT2 HEIGHT3
0         0130    0504
1         0200    0505
2         0142    0410
3         0190    0510
4         7777    0503
5         0172    0502
6         0150    0505
7         0130    0505
8         0205    0601
9         0151    0503
10        0199    0600
11        0170    0503
12        0160    0503
13        0200    0504
14        0285    0511
15        0217    0506
16        0230    0509
17        0250    0601
18        0155    0504
19        0138    0504
20        0240    0507
21        0240    0504
22        0185    0505
23        0165    0509
24        0131    0504
25        0119    0502
26        0180    0505
27        0169    0506
28        0128    0506
29        0218    0603
...        ...     ...
437406     NaN     NaN
437407    0320    0603
437408    0125    0501
437409    0158    0505
437410    0155    0502
437411    0150    0506
437412    0138    0502
437413    0135    0506
437414    0178    0506
437415    0126    0506
437416    0115    0411
437417    0183    0503
437418    0153    0504
437419    0138    0500
437420    0260    0503
437421    0160    0507
437422    0200    0601
437423    0126    0500
437424    0340    0507
437425    0298    0600
437426    0160    0508
437427    0160    0506
437428    0163    0502
437429    0172    0508
437430    0170    0503
437431    0098    0502
437432    7777    0502
437433    0156    0500
437434    0137    0501
437435    0192    0510

[437436 rows x 2 columns]

The `loc()` and `iloc()` methods (see the table on p. 144-5 of *Python for Data Analysis*)

In [ ]:
#Get the first row.
brfss.loc[0]

In [ ]:
#Get WEIGHT2 from the third row.
brfss.loc[2, 'WEIGHT2']

In [ ]:
#Same thing, but using the integer index instead of the column name.
brfss.iloc[2,-2]

In [ ]:
brfss.at[2, 'WEIGHT2']

In [ ]:
brfss.iat[2, -2]

In [ ]:
brfss.loc[:100, ['HEIGHT3', 'WEIGHT2']]

In [7]:
#Find rows based on a value
brfss[brfss['WEIGHT2']=='9999']

state  imonth  iday  iyear  dispcode  genhlth  physhlth  menthlth  \
61          1       2     1   2018      1100      1.0      88.0      88.0   
69          1       2     1   2018      1100      4.0      30.0      88.0   
90          1       2     3   2018      1100      1.0      88.0      14.0   
118         1       3     5   2018      1100      2.0      88.0      88.0   
129         1       3    15   2018      1100      1.0      88.0      88.0   
137         1       1    13   2018      1100      2.0      88.0      88.0   
159         1       1    11   2018      1100      2.0      88.0      88.0   
162         1       1     8   2018      1100      3.0      88.0      14.0   
172         1       1    25   2018      1100      2.0      88.0      88.0   
191         1       2     3   2018      1100      4.0      30.0      88.0   
231         1       3     1   2018      1100      5.0      30.0      15.0   
234         1       3    12   2018      1100      4.0      20.0      10.0   
271         1       1    11   2018      1100      3.0      88.0      88.0   
296         1       1     8   2018      1100      3.0      88.0      88.0   
320         1       3     1   2018      1100      2.0      10.0      88.0   
324         1       2     2   2018      1200      2.0      88.0      88.0   
347         1       3    24   2018      1100      2.0      88.0      88.0   
465         1       3     2   2018      1100      4.0      15.0      88.0   
475         1       3    22   2018      1100      3.0      88.0      88.0   
529         1       3     6   2018      1200      3.0      99.0      88.0   
534         1       2    14   2018      1100      5.0      30.0      88.0   
573         1       3    28   2018      1100      2.0      88.0      88.0   
590         1       1    11   2018      1100      3.0       3.0       7.0   
678         1       3     2   2018      1100      3.0      88.0      88.0   
694         1       1    23   2018      1200      2.0       7.0       5.0   
719         1       2    27   2018      1100      5.0      20.0      15.0   
728         1       2    15   2018      1100      1.0      88.0      88.0   
744         1       3     1   2018      1100      3.0      15.0      88.0   
773         1       3    22   2018      1100      2.0      88.0      88.0   
777         1       1    22   2018      1100      3.0      88.0      88.0   
...       ...     ...   ...    ...       ...      ...       ...       ...   
431704      6      11     5   2018      1100      1.0      10.0      88.0   
431715      6      11     1   2018      1200      3.0      88.0      88.0   
431766      6      12    10   2018      1100      3.0      88.0      88.0   
431803      6      11    20   2018      1100      4.0      99.0      99.0   
431808      6      12    10   2018      1200      1.0       7.0      88.0   
431814      6      12     6   2018      1100      2.0      88.0      88.0   
431962      6      10    23   2018      1100      4.0      88.0      88.0   
432005      6      12     3   2018      1200      1.0      88.0      88.0   
432257      6       5    31   2018      1200      3.0       1.0       5.0   
432272      6       5    12   2018      1100      4.0      88.0      88.0   
432276      6       5     7   2018      1100      2.0      88.0       1.0   
432319      6       6     9   2018      1100      3.0      88.0      88.0   
432404      6      10     8   2018      1200      3.0      88.0      88.0   
432441      6      10    10   2018      1200      1.0      88.0      88.0   
432466      6      11     5   2018      1100      3.0      88.0      88.0   
432709      2       5    12   2018      1100      2.0      88.0       2.0   
432987      2       8    21   2018      1100      3.0      15.0      25.0   
433173      2       1    12   2019      1100      3.0       3.0      88.0   
433642      2       3     8   2018      1100      2.0      88.0      88.0   
434574      2       8     6   2018      1100      2.0      88.0      30.0   
4346

The [`shape()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.shape.html) method gives you the height and width of your DataFrame.

In [ ]:
print(brfss.shape)
#Drop any rows without weight
brfss.dropna(subset=['WEIGHT2'], inplace=True)
print(brfss.shape)

https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.isin.html

In [12]:
brfss[np.isin(brfss['WEIGHT2'], ['7777', '9999'], invert=True)].shape

(415324, 15)

#### Derived Fields

In [10]:
brfss = brfss[np.isin(brfss['WEIGHT2'], ['7777', '9999'], invert=True)]
brfss = brfss[~brfss.WEIGHT2.str.startswith('1')]

TypeError: bad operand type for unary ~: 'float'

In [11]:
brfss['wtunit'] = brfss.WEIGHT2.str[0].astype(np.uint8)
brfss['wt'] = brfss.WEIGHT2.str[1:].astype(np.uint16)
brfss

ValueError: cannot convert float NaN to integer

[`np.where()`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.where.html) provides the equivalent of an *if-then-else* statement on each observation in a DataFrame:

In [ ]:
lbsperkg = 2.205
brfss['wtlbs'] = np.where(brfss.wtunit==9, brfss.wt*lbsperkg, brfss.wt).astype(np.int16)

In [ ]:
brfss[brfss.wtunit==9]

##### Practice

Create a `htinches` column from the `brfss.HEIGHT3` column. 

1. Remove rows where `HEIGHT3` is 7777 ("Don't know/Not sure"),  9999 ("Refused"), or NaN.
2. If the first character of `HEIGHT3` is '9', multiply the remaining three digits by `cmtoin` (defined below) to get height in inches.
3. If the first character of `HEIGHT3` is '0', the second character is feet, and the third and fourth are inches ('0601' means six feet, one inch). Convert this to inches.

See page 36 of the codebook for details on the `HEIGHT3` field.

In [8]:
cmtoin = 0.3937

#### Summary statistics and aggregation

In [9]:
brfss.groupby(['persdoc2', 'poorhlth']).size() #or .mean()

persdoc2  poorhlth
1.0       1.0          8301
          2.0          9481
          3.0          6023
          4.0          3408
          5.0          6584
          6.0          1075
          7.0          3308
          8.0           811
          9.0           189
          10.0         5593
          11.0           66
          12.0          524
          13.0           65
          14.0         1754
          15.0         5822
          16.0          141
          17.0           91
          18.0          169
          19.0           19
          20.0         3754
          21.0          487
          22.0           70
          23.0           46
          24.0           75
          25.0         1438
          26.0           71
          27.0           95
          28.0          288
          29.0          108
          30.0        15053
                      ...  
7.0       15.0           24
          18.0            2
          20.0           16
          21.0            3
 

In [ ]:
# | means 'or'
# ph=1 if you were sick more than 5 days, 0 otherwise:
brfss['ph'] = np.where((brfss['poorhlth'] > 30) | (brfss['poorhlth'] <= 5) | (brfss['poorhlth'].isnull()), 0, 1)
brfss

In [ ]:
#brfss.groupby(['persdoc2', 'ph']).describe()
brfss.ph.groupby(brfss.persdoc2).size()

In [ ]:
#Calculate percentages instead of raw numbers.
docph = brfss.groupby(['persdoc2', 'ph']).size()
docph.groupby(level=0).apply(lambda x: 100 * x / float(x.sum()))

We can bin or categorize numeric variables with [`pd.cut()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.cut.html).

In [ ]:
## 88 means 'none'; want to bin it separately from "Don't know" and "refused"
brfss.loc[brfss.poorhlth==88, 'poorhlth']=51
bins = [1, 5, 10, 20, 30, 51, 70]
brfss['phcats'] = pd.cut(brfss.poorhlth, bins, 
                         labels=['less than 5', 'less than 10', 
                                 'less than 20', 'more than 20', 'none', "don't know/refused"])
brfss

##### Practice

Create a new column that divides `iday` into 3 bins (1-10, 11-20, 21+).

2) Group `brfss` by `hlthpln1` and `medcost` and create a table like the one above (for `persdoc2` and `ph`) with percentages for each subgroup.